### ISSUE TITLE
Editable grid as popup cell editor close unexpectedly in 21.0.1

### ISSUE THREAD

**Issue by weihe8892** [ag-grid/ag-grid#3212]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of ag-Grid'
[] support request => see 'Requesting Community Support'
```

**Customers of ag-Grid**

If you are a customer you are entitled to use the ag-Grid's customer support system (powered by Zendesk). Please use that channel for guaranteed response from the ag-Grid team with regards bugs, feature requests and support.

**Requesting Community Support**

If you are not a customer of ag-Grid, ag-grid staff will label your issue as managed-by-the-community. This means that ag-Grid staff is not going to be actively looking into it and it will get closed if inactive for more than one month. The community is welcome to help with this question/support issue.

**Providing a Reproducible Scenario**
Accepted reproducible scenarios are
- A description of the detailed steps to reproduce your behaviour in one of our examples in the docs.
- A plunker
     1) Working fine in version 20.2.0 
           https://embed.plnkr.co/nsB6naLnTB7gqJKjrQRH/
 
   2) Break in 21.0.1
          https://embed.plnkr.co/g1NMBRUWwW6lhMY2Plxi/

If you decide to send us a plunkr, from any example in our website use the plunkr button in there to fork your own code by following the steps below:

- Select the framework that is appropriate to you from the drop-down
- Open it in plunker. (Use the button plunker in our example)
- Add your changes so that the behaviour is reproduced
- Save and Freeze the plunker(On the top left corner)
- Send us the link to the plunker(You can copy the URL from the browser)

If reporting a bug make sure to state.

Current behaviour.  

Expected behaviour. If possible back this up with our docs/examples if possible

**Current behavior**

      -- We have an editable grid as popup cell editor, in which has a ng-select component as cell editor. In ag-grid 21.0.1, after select an item, popup editor close unexpectedly and  no event was received.

![ag-grid-21 0 1](https://user-images.githubusercontent.com/6708815/59815211-e8a7fc00-92e5-11e9-8eef-1f70fc42a5ce.gif)

**Expected behavior**

    -- in version 20.2.0, after select an item, popup editor stays open and selection event was received.
![ag-grid-20 2 0](https://user-images.githubusercontent.com/6708815/59815227-ee9ddd00-92e5-11e9-97bf-1f6efb423d46.gif)
**Please tell us about your environment:**

   -- Angular version 7.2.15 
   -- Angular cli 7.3.9
* **ag-Grid version:** X.X.X

* **Browser:**

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]

**Comment by NikitaIT** [comment]
I updated 20-> 21 and all my element-ui drop-down elements, when pressed, began to close cell-editor.
![image](https://user-images.githubusercontent.com/16810067/60264024-e3047480-98ea-11e9-8f55-7107bb8d2069.png)
I would like an explanation of this change in the file.
ag-grid-community/dist/lib/widgets/popupService.js

**Comment by NikitaIT** [comment]
@weihe8892 Quick bug fix, works for me.

```js
// main.js
import { PopupService } from 'ag-grid-community';
import { getAddPopup } from './getAddPopup';
(function fixPopupInCellEdditors () {
  PopupService.prototype.addPopup = getAddPopup('click')
})();
```
```ts
// getAddPopup.ts
import { Utils as _, PopupService, Constants, Events } from "ag-grid-community";

export function getAddPopup(eventName: string = "mousedown") {
  return function addPopup(
    this: PopupService,
    modal: boolean,
    eChild: any,
    closeOnEsc: boolean,
    closedCallback?: () => void,
    click?: MouseEvent | Touch | null
  ): (event?: any) => void {
    // @ts-ignore
    const eDocument = this.gridOptionsWrapper.getDocument();

    if (!eDocument) {
      console.warn("ag-grid: could not find the document, document is empty");
      return () => {};
    }

    eChild.style.top = "0px";
    eChild.style.left = "0px";
    // @ts-ignore
    const ePopupParent = this.getPopupParent();

    const popupAlreadyShown = _.isVisible(eChild);
    if (popupAlreadyShown && ePopupParent.contains(eChild)) {
      return () => {};
    }

    // add env CSS class to child, in case user provided a popup parent, which means
    // theme class may be missing
    const eWrapper = document.createElement("div");
    // @ts-ignore
    const theme = this.environment.getTheme();

    if (theme) {
      _.addCssClass(eWrapper, theme);
    }

    _.addCssClass(eWrapper, "ag-popup");
    _.addCssClass(
      eChild,
      // @ts-ignore
      this.gridOptionsWrapper.isEnableRtl() ? "ag-rtl" : "ag-ltr"
    );

    eWrapper.appendChild(eChild);

    ePopupParent.appendChild(eWrapper);
    // @ts-ignore
    this.activePopupElements.push(eChild);

    let popupHidden = false;

    const hidePopupOnKeyboardEvent = (event: KeyboardEvent) => {
      const key = event.which || event.keyCode;
      if (
        key === Constants.KEY_ESCAPE &&
        eWrapper.contains(document.activeElement)
      ) {
        hidePopup(null);
      }
    };

    const hidePopupOnMouseEvent = (event: MouseEvent) => {
      hidePopup(event);
    };

    const hidePopupOnTouchEvent = (event: TouchEvent) => {
      hidePopup(null, event);
    };

    const hidePopup = (
      mouseEvent?: MouseEvent | null,
      touchEvent?: TouchEvent
    ) => {
      // we don't hide popup if the event was on the child, or any
      // children of this child
      // @ts-ignore
      if (this.isEventFromCurrentPopup(mouseEvent, touchEvent, eChild)) {
        return;
      }

      // if the event to close is actually the open event, then ignore it
      // @ts-ignore
      if (this.isEventSameChainAsOriginalEvent(click, mouseEvent, touchEvent)) {
        return;
      }

      // this method should only be called once. the client can have different
      // paths, each one wanting to close, so this method may be called multiple times.
      if (popupHidden) {
        return;
      }
      popupHidden = true;

      ePopupParent.removeChild(eWrapper);
      // @ts-ignore
      _.removeFromArray(this.activePopupElements, eChild);

      eDocument.removeEventListener("keydown", hidePopupOnKeyboardEvent);
      eDocument.removeEventListener(eventName, hidePopupOnMouseEvent);
      eDocument.removeEventListener("touchstart", hidePopupOnTouchEvent);
      eDocument.removeEventListener("contextmenu", hidePopupOnMouseEvent);
      // @ts-ignore
      this.eventService.removeEventListener(
        Events.EVENT_DRAG_STARTED,
        hidePopupOnMouseEvent
      );
      if (closedCallback) {
        closedCallback();
      }
    };

    // if we add these listeners now, then the current mouse
    // click will be included, which we don't want
    window.setTimeout(() => {
      if (closeOnEsc) {
        eDocument.addEventListener("keydown", hidePopupOnKeyboardEvent);
      }
      if (modal) {
        eDocument.addEventListener(eventName, hidePopupOnMouseEvent);
        // @ts-ignore
        this.eventService.addEventListener(
          Events.EVENT_DRAG_STARTED,
          hidePopupOnMouseEvent
        );
        eDocument.addEventListener("touchstart", hidePopupOnTouchEvent);
        eDocument.addEventListener("contextmenu", hidePopupOnMouseEvent);
      }
    }, 0);

    return hidePopup;
  };
}
```

**Comment by weihe8892** [comment]
Thanks @NikitaIT

**Comment by jessorlisa** [comment]
I found another simpler approach: Add the `ag-custom-component-popup` class to the ng-select component.

More details can be found here: https://www.ag-grid.com/javascript-grid-filter-component/#custom-filters-containing-a-popup-element

**Comment by davis4acca** [comment]
Hi,

Your ticket has been flagged as in_zendesk, it means that we recognize this as an issue worth investigating so we have moved it to our official support channel for customers (zendesk)

We will carry on there with the investigation and we will update you as soon as possible.

Thanks

**Comment by NikitaIT** [comment]
> I found another simpler approach: Add the `ag-custom-component-popup` class to the ng-select component.
> 
> More details can be found here: https://www.ag-grid.com/javascript-grid-filter-component/#custom-filters-containing-a-popup-element

Thanks, but this is not always possible and requires many changes to the working code.

@davis4acca I think you should put this information in a more prominent place in the documentation because it is needed for the most popular control - drop-down lists, usually using popper js. And now she is in an unobvious place.

**Comment by yannickvandervleuten** [comment]
I have tried using both `(click)="$event.stopPropagation()"` and  `ag-custom-component-popup`. But neither seems to work. I've tried placing them on multiple different elements but it doesn't affect it. 

https://angular-wvc3vv.stackblitz.io 
See stackblitz for example of project with `event.stopPropagation()`.

What did work however was @weihe8892's dirty fix. Which I rewrote a bit to fit the new version.
